### PR TITLE
FlaxGPT2

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?

[From Patrick:]

This PR adds GPT2 for Flax and lays the groundwork for `generation` in Flax. 
The model is a straightforward translation from PyTorch with two exceptions:

1.) the `past_key_values/caching` is more complex since `generation` has to compile to be usable on TPU with Flax/Jax. This means that the cache cannot grow dynamically, but has to be a static, padded tensor that is filled during generation. This means that the cache can now only be used as follows:
i) decide on a `max_length` to which the model should generate at most and call `model.init_cache(batch_size, max_length)`. This will then add a `"cache"` variable to the expected variables by the model and returns a 0 padded tensor of shape `(batch_size, max_length)`. 
ii) This cache is then subsequently passed & retrieved similar to how it's done in PyTorch:  `logits, past_key_values = model(input_ids, past_key_values=cache)`. When doing so, internally, the `cache` is marked as `mutable` so that the Flax model can step-by-step fill the padded tensor. We also need to pass a `_use_cache` boolean internally to tell the model to initialize the cache in the first place.
iii) The implemented caching system differs slightly from the original flax caching system [here](https://github.com/google/flax/blob/6111a4b2c5cb56520549c6dfc911135fbd8dbb7f/flax/linen/attention.py#L252) since we want to allow the very first decoding pass to take an `input_prompt` that is longer than 1 (*e.g.* for GPT3-like few-shot prompting) - see: https://github.com/google/flax/issues/1317.
iv) A test verifies that compiled generation works as expected
v) This also means that when using the cache, the `attention_mask` has to have the same static shape as the cache, I guess we could have better error messages here to, but I think this should mostly be handled by the generation method which will be added in a follow-up PR: https://github.com/huggingface/transformers/pull/11685.

2.) Since in Flax we rely on `flax.linen.dot_product_attention`, we are adding the attention mask once whereas in PyTorch we are setting certain values to the padding value via `torch.where(...)` [here](https://github.com/huggingface/transformers/blob/c73e35323d5805bff27c5dbd9a5691008be1316a/src/transformers/models/gpt2/modeling_gpt2.py#L187) and are adding large negative numbers [here](https://github.com/huggingface/transformers/blob/c73e35323d5805bff27c5dbd9a5691008be1316a/src/transformers/models/gpt2/modeling_gpt2.py#L191). This is unnecessarily costly in Flax and only matters for unrealistic cases where tokens in the middle of the sequence are padded. For all realistic cases PT and Flax behaves exactly the same, which is shown in the overwritten comparison tests.

### CODE REVIEW

File: src/transformers/models/gpt2/modeling_flax_gpt2.py [hunk 1]
```diff
@@ -0,0 +1,645 @@
...
+            else:
+                position_ids = jnp.broadcast_to(jnp.arange(sequence_length)[None, :], (batch_size, sequence_length))
+
+        if attention_mask is None:
+            # if past_key_values are passed we need to create an
+            # attention_mask of the same length as `cache_length`
+            if past_key_values is not None:
+                cache_length = next(
+                    iter(v for k, v in flatten_dict(unfreeze(past_key_values)).items() if k[-1] != "cache_index")
+                ).shape[1]
```
Comment (sgugger) [inline src/transformers/models/gpt2/modeling_flax_gpt2.py:402]: This is really a weird way of getting the cache length.

File: src/transformers/models/gpt2/modeling_flax_gpt2.py [hunk 2]
```diff
@@ -0,0 +1,645 @@
...
+            else:
+                cache_length = sequence_length
+
+            # Note that usually one would have to put 0's in the
+            # attention_mask for x > input_ids.shape[-1] and x < cache_length.
+            # But since GPT2 uses a causal mask, those positions are masked anyways.
+            # Thus we can create a single static attention_mask here,
+            # which is more efficient for compilation
```
Comment (sgugger) [inline src/transformers/models/gpt2/modeling_flax_gpt2.py:410]: Same nit as before for the comments.

File: src/transformers/models/gpt2/modeling_flax_gpt2.py [hunk 3]
```diff
@@ -0,0 +1,645 @@
...
+        if dropout_rng is not None:
+            rngs["dropout"] = dropout_rng
+
+        inputs = {"params": params or self.params}
+
+        # if past_key_values are passed then cache is already initialized
+        # a private flag _use_cache has to be passed down to
+        # ensure cache is used.
+        # It has to be made sure that cache is marked as mutable so that
+        # it can be changed by FlaxGPT2Attention module
```
Comment (sgugger) [inline src/transformers/models/gpt2/modeling_flax_gpt2.py:424]: Same nit as before for the comments.

File: src/transformers/models/gpt2/modeling_flax_gpt2.py [hunk 4]
```diff
@@ -0,0 +1,645 @@
...
+        self.blocks = [
+            FlaxGPT2Block(self.config, name=str(i), dtype=self.dtype) for i in range(self.config.num_hidden_layers)
+        ]
+
+    def __call__(
+        self,
+        hidden_states,
+        attention_mask=None,
+        deterministic: bool = True,
+        _use_cache: bool = False,
```
Comment (sgugger) [inline src/transformers/models/gpt2/modeling_flax_gpt2.py:473]: Same as before, why is this private?

File: src/transformers/models/gpt2/modeling_flax_gpt2.py [hunk 5]
```diff
@@ -0,0 +1,645 @@
...
+        self.h = FlaxGPT2BlockCollection(self.config, dtype=self.dtype)
+        self.ln_f = nn.LayerNorm(epsilon=self.config.layer_norm_epsilon, dtype=self.dtype)
+
+    def __call__(
+        self,
+        input_ids,
+        attention_mask,
+        position_ids,
+        deterministic=True,
+        _use_cache: bool = False,
```
Comment (sgugger) [inline src/transformers/models/gpt2/modeling_flax_gpt2.py:536]: Why is this private in the main API when it's public in PyTorch and TensorFlow?

File: tests/test_modeling_flax_common.py [hunk 6]
```diff
@@ -266,11 +266,15 @@ def model_jitted(input_ids, attention_mask=None, token_type_ids=None):
                     self.assertEqual(jitted_output.shape, output.shape)
 
                 @jax.jit
-                def model_jitted_return_dict(input_ids, attention_mask=None, token_type_ids=None):
+                def model_jitted_return_dict(
+                    input_ids,
+                    attention_mask=None,
+                    **kwargs,
+                ):
```
Comment (sgugger) [inline tests/test_modeling_flax_common.py:273]: There is no reason to add all those new lines: it fits in one.

File: tests/test_modeling_flax_common.py [hunk 7]
```diff
@@ -266,11 +266,15 @@ def model_jitted(input_ids, attention_mask=None, token_type_ids=None):
...
                     return model(
                         input_ids=input_ids,
                         attention_mask=attention_mask,
-                        token_type_ids=token_type_ids,
+                        **kwargs,
```
Comment (sgugger) [inline tests/test_modeling_flax_common.py:277]: Same, this model call should fit in one line.

File: tests/test_modeling_flax_gpt2.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+            eos_token_id=self.eos_token_id,
+            pad_token_id=self.pad_token_id,
+            gradient_checkpointing=gradient_checkpointing,
+        )
+
+        return (
+            config,
+            input_ids,
+            input_mask,
+        )
```
Comment (sgugger) [inline tests/test_modeling_flax_gpt2.py:110]:
```suggestion
        return (config, input_ids, input_mask)
```

**Review by sgugger — APPROVED** [review]
Great new addition!

File: src/transformers/models/gpt2/modeling_flax_gpt2.py [hunk 9]
```diff
@@ -0,0 +1,645 @@
...
+        self.attn = FlaxGPT2Attention(self.config, dtype=self.dtype)
+        self.ln_2 = nn.LayerNorm(epsilon=self.config.layer_norm_epsilon, dtype=self.dtype)
+        self.mlp = FlaxGPT2MLP(self.config, inner_dim, dtype=self.dtype)
+
+    def __call__(
+        self,
+        hidden_states,
+        attention_mask=None,
+        deterministic: bool = True,
+        _use_cache: bool = False,
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/transformers/models/gpt2/modeling_flax_gpt2.py:282]: True, changed it to `use_cache`

File: src/transformers/models/gpt2/modeling_flax_gpt2.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/models/gpt2/modeling_flax_gpt2.py:402]: Good point - I could also just take the first one tensor in the cache => that will be more readable

File: src/transformers/models/gpt2/modeling_flax_gpt2.py (hunk 5 above)
Comment (patrickvonplaten) [reply src/transformers/models/gpt2/modeling_flax_gpt2.py:536]: I think in Flax the user should never pass `use_cache` but handle everything over:

1) `init_cache(...)` and 
2) passing `past_key_values` or not

It's not possible to pass `use_cache` in the first pass to retrieve the cache like we do it in PyTorch, but we need to call `init_cache(...)` first. => I think it's better to afterward handle everything over `past_key_values` instead of making `use_cache` public. 
We still need the boolean though so that `init_cache(...)` can tell `FlaxGPT2Attention` to initialize the cache. 

=> Another option would be to instead of passing along `_use_cache`, we set a config parameter of `FlaxGPT2Attention` called `is_decoder` to `True` when `past_key_values` and to `False` when it is not instead of passing down a boolean flag & then have no `use_cache` at all? Is this maybe better? @sgugger

Comment (patrickvonplaten) [reply src/transformers/models/gpt2/modeling_flax_gpt2.py]: Actually that won't work either very cleanly since higher modules can't easily access lower modules as in PyTorch. I changed `_use_cache` to `init_cache` now to make a clear differentiation to PyTorch. Users never call `model.module.apply` directly anyways so that should be fine.

**Comment by patrickvonplaten** [comment]
@LysandreJik @sgugger - this PR is blocking a couple of other things. Let me know if you guys aren't really ok with some of those things
